wire-grpc-tests/src/test/java/com/squareup/wire/GrpcTest.kt

```diff
@@ -254,20 +250,100 @@ class GrpcTest {
     }
   }
 
-  // TODO(benoit) test before/after request completes, before after response completes?
   @Test
-  fun cancelDuplex() {
+  fun cancelDuplexBeforeRequestCompletes() {
     mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RouteChat"))
-    mockService.enqueue(Delay(500, TimeUnit.MILLISECONDS))
 
     val (requestChannel, responseChannel) = routeGuideService.RouteChat()
     runBlocking {
-      delay(200)
+      // We wait for mockService to process our actions.
+      delay(500)
+      responseChannel.cancel()
+      assertThat((requestChannel as Channel).isClosedForReceive).isTrue()
+      assertThat(requestChannel.isClosedForSend).isTrue()
+      mockService.awaitSuccess()
+      assertThat(callReference.get()?.isCanceled).isTrue()
+    }
+  }
+
+  @Test
+  fun cancelDuplexAfterRequestCompletes() {
+    mockService.enqueue(ReceiveCall("/routeguide.RouteGuide/RouteChat"))
+    mockService.enqueue(
+        ReceiveMessage(RouteGuideProto.RouteNote.newBuilder().setMessage("marco").build()))
+    mockService.enqueue(ReceiveComplete)
+
+    val (requestChannel, responseChannel) = routeGuideService.RouteChat()
+    runBlocking {
+      requestChannel.send(RouteNote(message = "marco"))
+      requestChannel.close()
+      // We wait for mockService to process our actions.
+      delay(500)
```

Comment:
I wonder if we can avoid this kind of delays and make the mock service suspend until it's processed the actions, is that possible?

Reply:
possibly? I find interleaving different parts of the code pretty difficult to do neatly